engine/apps/schedules/ical_utils.py

```diff
@@ -171,8 +177,14 @@ def get_shifts_dict(calendar, calendar_type, schedule, datetime_start, datetime_
         # Define on-call shift out of ical event that has the actual user
         if len(users) > 0 or with_empty_shifts:
             if type(event[ICAL_DATETIME_START].dt) == datetime.date:
-                start = event[ICAL_DATETIME_START].dt
-                end = event[ICAL_DATETIME_END].dt
+                if isinstance(schedule, OnCallScheduleWeb):
+                    start = max(event[ICAL_DATETIME_START].dt, event[ICAL_DATETIME_STAMP].dt.date())
+                    end = event[ICAL_DATETIME_END].dt
+                    if event.get(ICAL_RRULE, {}).get(ICAL_UNTIL):
+                        end = min(event[ICAL_DATETIME_END].dt, event[ICAL_RRULE][ICAL_UNTIL][0].date())
+                else:
+                    start = event[ICAL_DATETIME_START].dt
+                    end = event[ICAL_DATETIME_END].dt
```

Comment:
Wondering if this logic above could be refactored out somehow to use it here and reuse below?

Reply:
makes sense